File: leetcode-cn/53.最大子序和.py
#
# @lc app=leetcode.cn id=53 lang=python3
#
# [53] 最大子序和
#

# @lc code=start
from typing import List


class Solution:
    def maxSubArray(self, nums: List[int]) -> int:
        n = len(nums)
        dp = [0] * n
        dp[0] = nums[0]

        for i in range(1, n):
            dp[i] = max(dp[i - 1] + nums[i], nums[i])

        return max(dp)


# @lc code=end

File: leetcode-cn/C++/53.最大子序和.cpp
/*
 * @lc app=leetcode.cn id=53 lang=cpp
 *
 * [53] 最大子序和
 */

// @lc code=start
#include <vector>
using namespace std;

class Solution {
public:
    int maxSubArray(vector<int>& nums)
    {
        int ans = nums[0];

        int tmp = 0;

        for (int i = 0; i < nums.size(); i++) {
            tmp = max(nums[i], tmp + nums[i]);
            ans = max(ans, tmp);
        }

        return ans;
    }
};
// @lc code=end
